x-pack/plugins/data_visualizer/public/application/index_data_visualizer/index_data_visualizer.tsx

```diff
@@ -59,6 +89,54 @@ export const DataVisualizerUrlStateContextProvider: FC<
     null
   );
 
+  const [currentSessionId, setCurrentSessionId] = useState<string | undefined>(undefined);
+
+  useEffect(() => {
+    if (search.session) {
+      search.session.enableStorage({
+        getName: async () => {
+          // return the name you want to give the saved Search Session
+          return `Data visualizer`;
+        },
+        getLocatorData: async () => {
+          return {
+            id: DATA_VISUALIZER_APP_LOCATOR,
+            initialState: getLocatorParams({
+              ...services,
+              searchString,
+              indexPatternId: currentIndexPattern?.id,
+              savedSearchId: currentSavedSearch?.id,
+              shouldRestoreSearchSession: false,
+              searchSessionId: search.session.getSessionId(),
+            }),
+            restoreState: getLocatorParams({
+              ...services,
+              searchString,
+              indexPatternId: currentIndexPattern?.id,
+              savedSearchId: currentSavedSearch?.id,
+              shouldRestoreSearchSession: true,
+              searchSessionId: search.session.getSessionId(),
+            }),
+          };
+        },
+      });
+    }
+
+    const urlState = parseUrlState(searchString);
+
+    if (urlState.searchSessionId !== undefined && urlState.searchSessionId !== currentSessionId) {
+      search.session?.restore(urlState.searchSessionId);
+      setCurrentSessionId(urlState.searchSessionId);
+    } else {
+      const newSessionId = search.session?.start();
+      setCurrentSessionId(newSessionId);
+    }
+    return () => {
+      search.session.clear();
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
In this case this is because `searchString` includes all of the other dependencies. I added a comment here https://github.com/elastic/kibana/pull/121053/commits/a2359f1bc1e9630f67965f07fc3b7f667dfc0ba4 to clarify.